### DISCUSSION TITLE
USB Device DHCP (l4tbr0)

### DISCUSSION THREAD

**Discussion by len0rd** [OE4T/meta-tegra#815] [Q&A]
I'm trying to get to the bottom of an issue I see migrating from the Ubuntu 18.04 l4t release to a custom meta-tegra based build. Curious if anyone has suggestions of what I can try or investigate:

I'm connecting from a NX via the usb -> eth adapter to my dev machine. On Ubuntu the interfaces DHCP works as expected - assigning my dev-machine to 192.168.55.100. I can disconnect and reconnect the cable and the address gets re-assigned like normal.

With the meta-tegra image, the dev-machine only receives an ip the first time a cable is attached. If I disconnect/reconnect the cable my dev-machine recognizes the interface but is never assigned an IP.

Poking around both the images, I noticed meta-tegra uses networkd's dhcp and configures it in `recipes-bsp/l4t-usb-device-mode/l4t-usb-device-mode/70-l4tbr0.network`. On the ubuntu image, dhcpd is started via the service `nv-l4t-usb-device-mode-runtime`, The scripts for this service are located in `/opt/nvidia/l4t-usb-device-mode`. dhcpd is started/stopped on udev rule updates (see `/opt/nvidia/l4t-usb-device-mode/nv-l4t-usb-device-mode-state-change.sh`). Per the comments in `/opt/nvidia/l4t-usb-device-mode/nv-l4t-usb-device-mode-runtime-start.sh`, its done this way because dhcpd wont start if the interface is down. Does this relate to the issue I'm seeing on the meta-tegra image? Or am i missing something obvious? Suggestions for how to further diagnose?

Thank you for the help!

**Comment by madisongh** [comment]
I don't regularly use the USB device mode, but I just fired it up on a Xavier NX dev kit to try this out, and it seemed to work OK - I started a ping from my development host (which is running Ubuntu 21.04) to the NX, pulled the cable, waited a couple of seconds, then reconnected.  The ping paused while the connection was broken, of course, but resumed after reconnecting.

You can get more information about what systemd-networkd is doing on the device by adding `Environment=SYSTEMD_LOG_LEVEL=debug` to the `systemd-networkd.service` file, then rebooting and using `journalctl -b -u systemd-networkd -f` to watch its logs.
